Abstract away netty.2

scottweaver opened the pull request.

scottweaver committed:
Abstract Away Netty

* 'zio.http.netty' package created as the new home of all Netty-related implementations and artifacts.
* 'ChannelFuture' renamed 'NettyFutureExecutor' to better describe its purpose.
* Moved all Netty-specific implementation out of 'zio.http.Server' and into the new 'NettyDriver'
* Netty 'ChannelInitializer' and 'RequestHandler' are no longer hard coded and now can be provided using layers to the 'NettyDriver'
* Eliminated the somewhat confusing trend of using Netty names for zio http things.
* Eliminated a whole host of type aliases to make things a bit more clear.
* 'ServerChannelFactory' and 'ChannelFactory' consolidated to a single 'ChannelFactories' object.  Code has been streamlined as well.
  * It provides convenience layers for materializing both Server-specific and Client-specific into the environment by way of a 'ChannelType.Config'.
* 'EventLoopGroups' replaces 'EventLoopGroup' and provides a convenience layer similar to the ones mentioned above for ChannelFactory.
* Reduced API foot print by privatizing non-essential methods.
* Converted implicit arguments to explicit ones methods where they provided no real benefit by being implicit.  In many cases, they we being explicitly provided anyway.
* NettyRuntime replaces HttpRuntime.  'Strategy' removed in favor just implementing the abstract 'runtime' method on 'NettyRuntime'
  * 'sticky' renamed to 'usingSharedThreadPool' as it better reflects the semantics of the runtime.
  * 'default' renamed to 'usingDedicatedTrheadPool' as it better reflects the semantics of the runtime.
* 'Channel.Config' added so that there is no need for 'fromConfig' methods 'ChannelFactories' for both 'ClientConfig' and 'ServerConfig'.
* Many of the "unsafe" operations on 'ChannelHandlerContext' spread out all over the place now reside the 'ChannelHandlerContextOps' extension class.
* All client-specific Netty bits have made their way into 'zio.http.netty.client'.
* All server-specific Netty bits have made their way into 'zio.http.netty.server'.
* The 'R' type argument has been removed from all Netty internal type signatures as it served no real purpose.

scottweaver committed:
Abstract Away Netty

* 'zio.http.netty' package created as the new home of all Netty-related implementations and artifacts.
* 'ChannelFuture' renamed 'NettyFutureExecutor' to better describe its purpose.
* Moved all Netty-specific implementation out of 'zio.http.Server' and into the new 'NettyDriver'
* Netty 'ChannelInitializer' and 'RequestHandler' are no longer hard coded and now can be provided using layers to the 'NettyDriver'
* Eliminated the somewhat confusing trend of using Netty names for zio http things.
* Eliminated a whole host of type aliases to make things a bit more clear.
* 'ServerChannelFactory' and 'ChannelFactory' consolidated to a single 'ChannelFactories' object.  Code has been streamlined as well.
  * It provides convenience layers for materializing both Server-specific and Client-specific into the environment by way of a 'ChannelType.Config'.
* 'EventLoopGroups' replaces 'EventLoopGroup' and provides a convenience layer similar to the ones mentioned above for ChannelFactory.
* Reduced API foot print by privatizing non-essential methods.
* Converted implicit arguments to explicit ones methods where they provided no real benefit by being implicit.  In many cases, they we being explicitly provided anyway.
* NettyRuntime replaces HttpRuntime.  'Strategy' removed in favor just implementing the abstract 'runtime' method on 'NettyRuntime'
  * 'sticky' renamed to 'usingSharedThreadPool' as it better reflects the semantics of the runtime.
  * 'default' renamed to 'usingDedicatedTrheadPool' as it better reflects the semantics of the runtime.
* 'Channel.Config' added so that there is no need for 'fromConfig' methods 'ChannelFactories' for both 'ClientConfig' and 'ServerConfig'.
* Many of the "unsafe" operations on 'ChannelHandlerContext' spread out all over the place now reside the 'ChannelHandlerContextOps' extension class.
* All client-specific Netty bits have made their way into 'zio.http.netty.client'.
* All server-specific Netty bits have made their way into 'zio.http.netty.server'.
* The 'R' type argument has been removed from all Netty internal type signatures as it served no real purpose.

scottweaver committed:
Remove 'shrinks'

scottweaver committed:
Make AtomicReferences more palatable.
* AtomicReferences moved in the newly-minted 'Driver.Context'.

scottweaver committed:
Fix Scala 3 compilation issues

scottweaver committed:
Request stripped of all Netty artifacts and simplified.
* Remote address access moved from 'Http' to 'Request'

scottweaver committed:
Abstract Away Netty WIP

scottweaver committed:
[AAN] WIP Version now free and clear.

scottweaver committed:
Cleanup after rebase.

scottweaver committed:
Remove commented-out code

scottweaver committed:
Added spec for 'Versions'.

scottweaver committed:
Rebase from main

scottweaver committed:
Post-rebase changes.

jdegoes commented on zio-http/src/main/scala/zio/http/Driver.scala:
```diff
@@ -2,10 +2,35 @@ package zio.http
+
+object Driver {
+
+  private[zio] final case class Context private (
```
This class (`Driver.Context`) should be regarded as a private implementation detail for _some_ drivers, but not others, and moved to some place Netty-specific.

scottweaver replied:
It actually shouldn't exist at all.  It was replaced by new methods in the Driver.

jdegoes commented on zio-http/src/main/scala/zio/http/Http.scala:
```diff
@@ -723,10 +722,10 @@ object Http {
    */
-  def context: Http[Any, Nothing, Request, ChannelHandlerContext] =
-    Http.fromFunctionZIO[Request](request => ZIO.succeedUnsafe { implicit u => request.unsafe.context })
+  // def context: Http[Any, Nothing, Request, ChannelHandlerContext] = ???
```
Dead code

jdegoes commented on zio-http/src/main/scala/zio/http/Http.scala:
```diff
@@ -981,15 +980,16 @@ object Http {
-        case _                    => Http.fail(new IOException("Unable to get remote address"))
-      }
-    }
+  // def remoteAddress: Http[Any, IOException, Request, InetAddress] =
```
Dead code

jdegoes commented on zio-http/src/main/scala/zio/http/Http.scala:
```diff
@@ -1037,10 +1037,10 @@ object Http {
+   * using the request's ChannelHandlerContext.
    */
-  def usingContext[R, E](f: ChannelHandlerContext => HttpApp[R, E]): HttpApp[R, E] = context.flatMap(f(_))
+  // def usingContext[R, E](f: ChannelHandlerContext => HttpApp[R, E]): HttpApp[R, E] = context.flatMap(f(_))
```
Dead code

jdegoes commented on zio-http/src/main/scala/zio/http/Http.scala:
```diff
@@ -1037,10 +1037,10 @@ object Http {
   /**
-   * Provides low level access to an HttpApp to perform unsafe operations using
-   * the request's ChannelHandlerContext.
+   * [1453] Provides low level access to an HttpApp to perform unsafe operations
```
I don't know what these `[1453]` comments are about (here and elsewhere).

scottweaver replied:
Oops, they were there as a reference point.  That is the ticket number.

jdegoes commented on zio-http/src/main/scala/zio/http/Request.scala:
```diff
@@ -1,199 +1,55 @@
-   * Updates the headers using the provided function
-   */
-  final override def updateHeaders(update: Headers => Headers): Request = self.copy(headers = update(self.headers))
+final case class Request(
```
This change will need to be benchmarked since one of the reasons `Request` was a trait was to provide the ability to virtualize some of its fields, rather than be forced to materialize the body, header, method, url, version, and remote address (including its decoding and allocation cost).

Now, I am not sure if this was actually utilized anymore, or if the `trait` was more "forward thinking". We need to benchmark to make sure no regression here.

To accomodate high-performance use-cases, if necessary, we could introduce a `trait RequestLike` that defines all the same fields, and then read-only use cases could require `RequestLike` rather than `Request`.

scottweaver replied:
I assume your comment further down allays some the concern here?  Still, I will be running benchmarks.

jdegoes commented on zio-http/src/main/scala/zio/http/Request.scala:
```diff
@@ -1,199 +1,55 @@
-   * Remove trailing slash from path.
-   */
-  final def dropTrailingSlash: Request = self.copy(url = self.url.dropTrailingSlash)
+  def withMethod(method: Method): Request = self.copy(method = method)
```
I think `copy` is fine and we should probably delete all the `withXYZ` methods.

jdegoes commented on zio-http/src/main/scala/zio/http/Request.scala:
```diff
@@ -1,199 +1,55 @@
+   * Constructor for Request that provides sane defaults.
    */
-  def apply(
+  def make(
```
It's an anti-pattern (in my view) for a straightforward case class to have a `make` constructor. Rather, the default case class constructor should always be used: `Request(...)` (ordinary Scala) instead of `Request.make(...)` (weird Scala).

scottweaver replied:
I try to avoid default arguments in the base class itself, that's what I created a `make` so as to support sane defaults.

At previous gigs, we considered default values in a case class an anti-pattern.

jdegoes replied:
I also agree, but let's do that via `Request.default.copy(...)`.

jdegoes replied:
Or maybe `Request.get` would be more appropriate. Then a user could do: `Request.get.copy(...)`.

jdegoes commented on zio-http/src/main/scala/zio/http/Request.scala:
```diff
@@ -1,199 +1,55 @@
-    new Request {
-      override final def headers: Headers = Headers.make(jReq.headers())
-      override final def method: Method   = Method.fromHttpMethod(jReq.method())
-      override final def url: URL         = URL.fromString(jReq.uri()).getOrElse(URL.empty)
```
Looking at these, I suspect the use of `trait` for `Request` was not buying any performance gains, because every field is created every time. In addition, you succeeded to delete `val unsafe` which has negative performance implications.

scottweaver replied:
Deleting unsafe from all the public APIs is one of my end-goals for this effort.

scottweaver committed:
[PR Tasks] Remove dead and obsolete code.

scottweaver committed:
POst-rebase fixup.

jdegoes commented:
@scottweaver Can you resolve conflicts on `|ServerConfig`?

scottweaver commented:
Rebased and pushed.

scottweaver committed:
Fixed a bug (my fault) that was introduced in 'Http#whenPathEq'

jdegoes commented on zio-http/src/main/scala/zio/http/netty/Versions.scala:
```diff
@@ -0,0 +1,21 @@
+      case _                    => throw new IllegalArgumentException(s"Unsupported HTTP version: $version")
+    }
+
+  def make(version: Version): HttpVersion = version match {
```
This can throw exceptions.

jdegoes replied:
Or maybe not.

scottweaver replied:
~`Versions#make(HttpVersion)` removed and all code now inlined into `SeverInboundHandler`.~

jdegoes commented on zio-http/src/main/scala/zio/http/netty/Versions.scala:
```diff
@@ -0,0 +1,21 @@
+
+object Versions {
+  import Version._
+  def make(version: HttpVersion)(implicit unsafe: Unsafe, trace: Trace): Version =
```
`make` is good for a constructor in the companion object of a data type, but for a utility like this, `convertVersion` is probably better.

jdegoes replied:
Maybe `convertNettyToZIO` and `convertZIOToNetty`, respectively?

scottweaver replied:
`Versions#make(HttpVersion)` removed and all code now inlined into `SeverInboundHandler`.

scottweaver committed:
Perf iteration 1
* Inline construction of ZIO Http Requests to ServerInboundHandler.
* Requests removed as it was only ever used in ServerInboundHandler.

scottweaver committed:
Perf iteration 2 'Versions.make' inlined to 'ServerInboundHandler'.
* 'Versions.make' deleted as it was only used by 'ServerInboundHandler'.

scottweaver committed:
Perf iteration 3.  Stop eagerly fetching the remote address.

scottweaver committed:
Perf iteration 4.  Try some inlining?

scottweaver committed:
Perf iteration 5.  Move all 'ChannelHandlerContextOps' logic to 'ServerInboundHandler'.

I am sure that usage of extension methods here was a bad choice in terms of performance impact.  It also appears that 100% of the operations in 'ChannelHandlerContextOps' were only ever used by 'ServerInboundHandler".

scottweaver committed:
Perf iterations 6.  Track down IOException: Connection Reset.

scottweaver committed:
Perf iteration 7.  Reduce debug output.

scottweaver committed:
Perf iteration 8. Explicitly evaluate exception.
* Also fixed test regression on 'isReadKey'.

scottweaver committed:
Fix unmoored docs

scottweaver committed:
Switch to 'zio/FrameworkBenchmarks'

scottweaver committed:
Correctly updated CIO actions to use 'zio/FrameworkBenchmarks'.

scottweaver committed:
Ran 'githubWorkflowGenerate'

scottweaver committed:
Perf iteration 9.  Log connection reset explicitly.
* 'Http#whenPathEq' is no longer unsafe.

scottweaver committed:
[DEBUG] Print contents of bench server

scottweaver committed:
Per iteration 10. Fixed issue with JSON benchmark output.
* The field in the benchmark must be named "message" not "greetings".

scottweaver committed:
Removed source code debug from CI

scottweaver committed:
Added missing comma to "hello, world"

scottweaver committed:
PR Tasks
* 'Body#empty' from 'def' to 'val'.
* 'Request#*' constructors refactored to encompass the most common use-cases.
*  Convenience methods removed from 'Request' in favor of just using one of new convenience methods + 'copy'.
*  'Versions#make()' renamed to 'Versions#convertToZIOToNetty()'.
*  Remove dead code.

scottweaver committed:
Added convenience script to run benchmarks locally.

scottweaver committed:
Deleted convenience methods off of 'Request'.

scottweaver committed:
Formatting
